Add initial embedded rust support

vladandrew opened the pull request:
This patch adds embedded rust support. Only the core crate is available. Internal libraries may be written in rust now.
Right now, the sources should have #![no_std] and a panic handler but future work will integrate this in a rust internal library.

vladandrew committed:
build: Add embedded rust support

This patch adds embedded rust support. Only the core crate
is available. Internal libraies may be written in rust now.
We're using rustc as a rust compiler. The compilation flags
are different from GCC and thus we introduce new flags for
rust sources such as as a -Copt_lvl="2". LTO is available
only when using the clang toolchain for C/C++ sources.
DCE is available and done at linking.
The sources should have #![no_std] and a panic handler, e.g:

```
use core::panic::PanicInfo;

fn panic(_info: &PanicInfo) -> ! {
    loop {}
}
```

Signed-off-by: Vlad-Andrei Badoiu <vlad_andrei.badoiu@upb.ro>

skuenzer commented on Makefile:
```diff
@@ -563,6 +565,7 @@ CC		:= $(CONFIG_CROSS_COMPILE)$(CONFIG_COMPILER)
 CPP		:= $(CC)
 CXX		:= $(CPP)
 GOC		:= $(CONFIG_CROSS_COMPILE)gccgo-7
+RUSTC		:= rustc
```
How would cross compiling work with rust? I would have expected something like
```Makefile
RUSTC		:= $(CONFIG_CROSS_COMPILE)rustc
```

vladandrew replied:
Rustc is using --target for specifying the target, I've modified the code accordingly.

skuenzer replied:
I think `$(CONFIG_CROSS_COMPILE)` includes a minus. We probably just need to remove the trailing `-` from the variable. However, do you know if the naming is exactly the same or do we need translating it? A tiny support script could do the job, e.g., `support/scripts/rustc-cross`

skuenzer commented on support/build/Makefile.rules:
```diff
@@ -540,6 +540,21 @@ $(eval $(call vprefix_lib,$(1),OBJS-y) += $(4))
+define buildrule_rs =
+$(4): $(2) | preprocess
+	$(call build_cmd,RUSTC,$(1),$(4),\
+		$(RUSTC) $$(RUSTCFLAGS) $$(RUSTCFLAGS-y) \
```
Can we also include the general compiler flags that we have, like the optimization flags, flags for ISR code and the file-variant specific flags?
Is `rustc` GCC flag compatible?

vladandrew replied:
Rustc is not GCC compatible, I've adapted the debug and optimization flags accordingly.

skuenzer replied:
Are there includes that need to be considered for `rustc`?
Can `rustc` generate something for dependency management?

skuenzer replied:
Can you also add `$$(RUSTCFLAGS_EXTRA)` to get the functionality as introduced with PR #255 ?

skuenzer reviewed (commented):
Hey, this looks really cool. I have some question regarding the integration and adoption of our flags that are set for the compilation units. Most importantly, DCE, LTO and ISR-safe code.

vladandrew commented:
Hi Simon, thank you for your review, I've solved all the comments and pushed the next version of the patch!

skuenzer commented on Makefile.uk:
```diff
@@ -21,6 +21,23 @@ CINCLUDES    += -I$(CONFIG_UK_BASE)/include
+# There is not level 3 in rustc
+RUSTCFLAGS-$(CONFIG_DEBUG_SYMBOLS_LVL3)     += -Cdebuginfo=2
+
+# Works only with clang
```
Please add this as a note.
```Makefile
# NOTE: rustc supports LTO only with clang
```

skuenzer commented on support/build/Makefile.rules:
```diff
@@ -540,6 +540,24 @@ $(eval $(call vprefix_lib,$(1),OBJS-y) += $(4))
 $(eval $(call vprefix_lib,$(1),CLEAN-y) += $(call build_clean,$(4)) $(call out2dep,$(4)))
 endef
 
+# We are not using most of the flags such as COMPFLAGS due to incompatibilities between rustc and GCC.
```
```Makefile
# NOTE: ...
```

vladandrew committed:
build: Add LLVM_TARGET_ARCH

Rustc uses the LLVM format for specifying the cross
compiler target: <arch><sub>-<vendor>-<sys>-<abi>.
This is different from GCC, in order to not have
to mantain a translation script, we provide the user
with an option to specify himself the target via
LLVM_TARGET_ARCH which is similar to CROSS_COMPILE.

Signed-off-by: Vlad-Andrei Badoiu <vlad_andrei.badoiu@upb.ro>

vladandrew commented:
Hi Simon, I've pushed a new version of this series. I've added a new patch for LLVM_TARGET_ARCH.

skuenzer commented on Makefile:
```diff
@@ -568,6 +570,13 @@ CC		:= $(CONFIG_CROSS_COMPILE)$(CONFIG_COMPILER)
+# We use rustc because the gcc frontend is experimental and missing features such
+# as borrowing checking
+ifneq ("$(origin CROSS_COMPILE)","undefined")
+RUSTC		:= rustc --target=$(CONFIG_CROSS_COMPILE)
```
Shouldn't this be set to `LLVM_TARGET_ARCH`?

vladandrew replied:
Solved :100:

skuenzer commented on Makefile:
```diff
@@ -515,6 +515,7 @@ ifeq ($(sub_make_exec), 1)
 ifeq ($(UK_HAVE_DOT_CONFIG),y)
 # Hide troublesome environment variables from sub processes
 unexport CONFIG_CROSS_COMPILE
+unexport LLVM_TARGET_ARCH
```
```diff
+unexport CONFIG_LLVM_TARGET_ARCH
```

vladandrew replied:
Done

skuenzer commented on Config.uk:
```diff
@@ -168,6 +168,14 @@ config CROSS_COMPILE
+		Same as running 'make LLVM_TARGET_ARCH=target' but stored for
+		default make runs in this build directory.  You don't
+		need to set this unless you want the configured build
+		directory to select the cross-compiler automatically.
```
Maybe add another sentence that this is used when compiling with LLVM compiles, like clang or rustc.

skuenzer reviewed (changes requested):
Can you add an extra flag called `$$(RUSTCFLAGS_EXTRA)` to get the functionality as introduced with PR #255 ?

skuenzer reviewed (approved):
Looks good to me. Thanks!

Reviewed-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

nderjung commented:
Hi,

@skuenzer can you re-approve this PR following @vladandrew new rebase?  I tried re-running the merge pipeline but it fails because it's re-using the state at of unikraft's `staging` branch of when you sent the original approval.

Thanks!

skuenzer reviewed (approved):
Looks good, thanks a lot!

Reviewed-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
